File: 1501~2000/1743. Restore the Array From Adjacent Pairs/main.py
"""
Source: https://rb.gy/nmjm6
Date: 2023/8/21
Skill:
Ref:
Runtime: 1055 ms, faster than 62.93%
Memory Usage: 66.47 MB, less than 67.96%
Time complexity:
Space complexity:
Constraints:
"""

import math, sys
from typing import List, Optional
from collections import defaultdict, Counter, deque
from heapq import heapify, heappush, heappop, nsmallest
import heapq
import functools
from bisect import bisect_left, bisect_right
from functools import cache


class Solution:
    def restoreArray(self, adjacentPairs: List[List[int]]) -> List[int]:
        adj_list, degree, n = defaultdict(list), defaultdict(int), len(adjacentPairs) + 1
        for a, b in adjacentPairs:
            adj_list[a].append(b)
            adj_list[b].append(a)
            degree[a] += 1
            degree[b] += 1

        q, left, right, is_left, res = deque(), 0, n - 1, True, [0 for _ in range(n)]
        for key, val in degree.items():
            if val == 1:
                q.append([key, is_left])
                is_left = not is_left
                degree[key] = 0

        while q:
            num, is_left = q.popleft()
            if is_left:
                res[left] = num
                left += 1
            else:
                res[right] = num
                right -= 1
            for adj in adj_list[num]:
                degree[adj] -= 1
                if degree[adj] == 1:
                    q.append([adj, is_left])
                    degree[adj] = 0

        return res


if __name__ == "__main__":
    s = Solution()

File: 1501~2000/1743. Restore the Array From Adjacent Pairs/main.cpp
/**
 * Source: t.ly/poCiS
 * Date: 2023/11/10
 * Skill:
 * Runtime: 316 ms, faster than 89.09% of C++ online submissions
 * Memory Usage: 106.47 MB, less than 81.21% of C++ online submissions
 * Time complexity: O(n)
 * Space complexity: O(n)
 * Constraints:
 *
 */


#include <vector>
#include <iostream>
#include <algorithm>
#include <map>
#include <set>
#include <unordered_set>
#include <queue>
#include <numeric>
#include <tuple>
#include <stack>
#include <bitset>
#include <functional>
#include <list>


using namespace std;


class Solution {
public:
    vector<int> restoreArray(vector<vector<int>>& adjacentPairs) {
        unordered_map<int, vector<int>> adj_list;
        for (auto &i: adjacentPairs) {
            adj_list[i[0]].push_back(i[1]);
            adj_list[i[1]].push_back(i[0]);
        }
        int cur, prev;
        for (auto it: adj_list) {
            if (it.second.size() == 1) {
                cur = it.first;
                break;
            }
        }
        vector<int> res;
        while (true) {
            prev = res.size() == 0 ? cur : res.back();
            res.push_back(cur);
            for (auto i: adj_list[cur]) {
                if (i != prev) {
                    cur = i;
                    break;
                }
            }
            if (cur == res.back()) break;
        }
        return res;
    }
};


int main() {
    Solution s;
    vector<vector<int>> arr{{4,8}, {2,8}};
    vector<int> nums{2,3,5};
}